formal-models/README.md

```diff
@@ -112,6 +112,215 @@ configuration:
 * [TLA⁺ specification](./dbftMultipool/dbftMultipool.tla)
 * [TLC Model Checker configuration](./dbftMultipool/dbftMultipool___AllGoodModel.launch)
 
+## Proposed dBFT 2.1 models
+
+Based on the liveness locks scenarios found by the TLC model checker in the
+[basic dBFT 2.0 model](#basic-dbft-20-model) we've developed two extensions of
+dBFT 2.0 protocol that allow to avoid the liveness lock problem and to preserve
+the safety properties of the algorithm. The extensions currently doesn't have
+the code-level implementation and presented as a TLA⁺ specifications ready to be
+reviewed and discussed. The improved protocol presented in the extensions will
+be referred below as dBFT 2.1.
+
+We've checked both dBFT 2.1 models with the TLC Model Checker against the same
+set of launch configurations that was used to reveal the liveness problems of the
+[basic dBFT 2.0 model](#basic-dbft-20-model). The improved models have larger
+set of states, thus, the TLC Model Checker wasn't able to finish the liveness
+requirements checks for *all* possible states. However, the checks have passed for
+a state graph diameter that was large enough to believe the presented models
+solve the dBFT 2.0 liveness lock problems.
+
+### dBFT 2.1 stages-based model
+
+The basic idea of this model is to split the consensus process into three subsequent
+stages marked as `I`, `II` and `III` at the scheme. To perform a transition between
+two subsequent stages each consensus node should wait for a certain set of messages
+to be received so that it's possible to gather a full picture of the neighbours'
+decisions in the current consensus round. At the same time, each of the stages has
+its own `ChangeView[1,2,3]` message to exit to the next consensus round (view) if
+something goes wrong in the current one and there's definitely no ability to
+continue consensus process in the current view. Below there's a short description
+of each stage. Please, refer to the model scheme and specification for further
+details.
+
+#### Stage I
+
+Once initialized, consensus node has two ways:
+1. Send its `PrepareRequest`/`PrepareResponse` message (and transmit to the
+  `prepareSent` state).
+2. Decide to go to the next view on timeout or any other valid reason like
+  transaction missing in the node's mempool via sending `ChangeView1` message
+  (and transmit to the `cv1` state).
+
+This scheme is quite similar to the basic dBFT 2.0 model except the new type of
+`ChangeView` message. After that the node enters stage `I` and waits for consensus
+messages of stage `I` (`PrepareRequest` or `PrepareResponse` or `ChangeView1`)
+from at least `M` neighbours which is needed to decide about the next actions.
+The set of received messages can be arranged in the following way:
+
+* `M` messages of `ChangeView1` type denote that `M` nodes have decided to change
+  their view directly after initialization due to invalid/missing `PrepareRequest`
+  which leads to immediate view changing.
```

Comment:
This is a "fail fast" route that is the same as with dbft 2.0 for the widespread case of missing primary. No additional delay is added, everything works as usual.